src/test/fuzz/util.h

```diff
@@ -512,4 +530,128 @@ void ReadFromStream(FuzzedDataProvider& fuzzed_data_provider, Stream& stream) no
+            ENOTCONN,
+            ENOTSOCK,
+            EOPNOTSUPP,
+            EPIPE,
+            EWOULDBLOCK,
+        };
+        if (m_fuzzed_data_provider.ConsumeBool()) {
+            return len;
+        }
+        const ssize_t r = m_fuzzed_data_provider.ConsumeIntegralInRange<ssize_t>(-1, len);
+        if (r == -1) {
+            SetFuzzedErrNo(m_fuzzed_data_provider, send_errnos);
+        }
+        return r;
+    }
+
+    ssize_t recv(char* buf, size_t len, int flags) override
+    {
+        const std::vector<int> recv_errnos{
+            EAGAIN,
+            EBADF,
+            ECONNREFUSED,
+            EFAULT,
+            EINTR,
+            EINVAL,
+            ENOMEM,
+            ENOTCONN,
+            ENOTSOCK,
+            EWOULDBLOCK,
+        };
+        assert(buf != nullptr || len == 0);
+        if (len == 0 || m_fuzzed_data_provider.ConsumeBool()) {
+            const ssize_t r = m_fuzzed_data_provider.ConsumeBool() ? 0 : -1;
+            if (r == -1) {
+                SetFuzzedErrNo(m_fuzzed_data_provider, recv_errnos);
+            }
+            return r;
+        }
+        const std::vector<uint8_t> random_bytes = m_fuzzed_data_provider.ConsumeBytes<uint8_t>(m_fuzzed_data_provider.ConsumeBool() ? len : m_fuzzed_data_provider.ConsumeIntegralInRange<size_t>(0, len));
+        if (random_bytes.empty()) {
+            const ssize_t r = m_fuzzed_data_provider.ConsumeBool() ? 0 : -1;
+            if (r == -1) {
+                SetFuzzedErrNo(m_fuzzed_data_provider, recv_errnos);
+            }
+            return r;
+        }
+        std::memcpy(buf, random_bytes.data(), random_bytes.size());
+        if (m_fuzzed_data_provider.ConsumeBool()) {
+            if (len > random_bytes.size()) {
+                std::memset(buf + random_bytes.size(), 0, len - random_bytes.size());
+            }
+            return len;
+        }
+        return random_bytes.size();
+    }
+
+#ifdef USE_POLL
+    int poll(short events, int timeout_ms) override
+    {
+        const std::vector<int> poll_errnos{
+            EFAULT,
+            EINTR,
+            EINVAL,
+            ENOMEM,
+        };
+        const int r = m_fuzzed_data_provider.ConsumeIntegralInRange<int>(-1, 1);
+        if (r == -1) {
+            SetFuzzedErrNo(m_fuzzed_data_provider, poll_errnos);
+        }
+        return r;
+    }
+#endif
+
+    int select(bool watch_read, bool watch_write, int timeout_ms) override
+    {
+        const std::vector<int> select_errnos{
+            EBADF,
+            EINTR,
+            EINVAL,
+            ENOMEM,
+        };
+        const int r = m_fuzzed_data_provider.ConsumeIntegralInRange<int>(-1, 1);
+        if (r == -1) {
+            SetFuzzedErrNo(m_fuzzed_data_provider, select_errnos);
+        }
+        return r;
+    }
+
+    bool IsSelectable() const override
+    {
+        return m_fuzzed_data_provider.ConsumeBool();
+    }
+};
+
+FuzzedSocket ConsumeSocket(FuzzedDataProvider& fuzzed_data_provider)
```

Comment:
Should use `inline`. See https://github.com/bitcoin/bitcoin/pull/20733

macOS build will warn with 

```
ld: warning: duplicate symbol 'ConsumeSocket(FuzzedDataProvider&)' in:
    test/fuzz/fuzz-addition_overflow.o
    libtest_fuzz.a(libtest_fuzz_a-util.o)
```

Reply:
Now addressed. Thanks!